solver/monomial_solve.c

```diff
@@ -248,43 +349,185 @@ int solve_mms_nd(spinor ** const Pup, spinor ** const Pdn,
     }
   } else
 #endif //TM_USE_QPHIX
-  if(solver_params->external_inverter == NO_EXT_INV){
-    if(solver_params->type==MIXEDCGMMSND){
-      if(usegpu_flag){
-	#ifdef HAVE_GPU      
-	  #ifdef TEMPORALGAUGE
-	    to_temporalgauge_mms(g_gauge_field , Qup, Qdn, Pup, Pdn, solver_params->no_shifts);
-	  #endif        
-	  iteration_count = dev_cg_mms_tm_nd(Pup, Pdn, Qup, Qdn, solver_params);  
-	  #ifdef TEMPORALGAUGE
-	    from_temporalgauge_mms(Qup, Qdn, Pup, Pdn, solver_params->no_shifts);
-	  #endif 
-	#endif
+  if(solver_params->type==MIXEDCGMMSND){
+    if(usegpu_flag){
+    #ifdef HAVE_GPU      
+      #ifdef TEMPORALGAUGE
+      to_temporalgauge_mms(g_gauge_field , Qup, Qdn, Pup, Pdn, solver_params->no_shifts);
+      #endif        
+      iteration_count = dev_cg_mms_tm_nd(Pup, Pdn, Qup, Qdn, solver_params);  
+      #ifdef TEMPORALGAUGE
+      from_temporalgauge_mms(Qup, Qdn, Pup, Pdn, solver_params->no_shifts);
+      #endif 
+    #endif
+    } else {
+      iteration_count = mixed_cg_mms_tm_nd(Pup, Pdn, Qup, Qdn, solver_params);
+    }
+  } else if (solver_params->type == CGMMSND){
+    iteration_count = cg_mms_tm_nd(Pup, Pdn, Qup, Qdn, solver_params);
+  }
+#ifdef DDalphaAMG
+  else if (solver_params->type == MG) {
+    // if the mg_mms_mass is larger than the smallest shift we use MG
+    if (mg_no_shifts > 0 || mg_mms_mass >= solver_params->shifts[0]) { 
+
+      int nshifts = solver_params->no_shifts;
+      int mg_nshifts = mg_no_shifts > nshifts ? nshifts:mg_no_shifts;
+      // if the mg_mms_mass is smaller than the larger shifts, we use CGMMS for those
+      // in case mg_no_shifts is used, then mg_mms_mass = 0
+      if(mg_mms_mass >= solver_params->shifts[0]) {
+        mg_nshifts = nshifts;
+        while (mg_mms_mass < solver_params->shifts[mg_nshifts-1]) { mg_nshifts--; }
+      }
+      // Number of initial guesses provided by gcmms
+      // README: tunable value. 1 it's fine for now.
+      int no_cgmms_init_guess = 2;
+      if(no_cgmms_init_guess > mg_nshifts) {
+        no_cgmms_init_guess = mg_nshifts;
       }
-      else{
-	iteration_count = mixed_cg_mms_tm_nd(Pup, Pdn, Qup, Qdn, solver_params);
+      if (mg_nshifts < nshifts) {
+        spinor ** Pup_cg = Pup+(mg_nshifts - no_cgmms_init_guess);
+        spinor ** Pdn_cg = Pdn+(mg_nshifts - no_cgmms_init_guess);
+        double * shifts_start = solver_params->shifts;
+        solver_params->no_shifts = nshifts - (mg_nshifts - no_cgmms_init_guess);
+        solver_params->shifts += (mg_nshifts - no_cgmms_init_guess);
+        solver_params-> type = CGMMSND;
+        if (no_cgmms_init_guess > 0) {
+          SWAP(solver_params->shifts[0], solver_params->shifts[no_cgmms_init_guess]);
+          SWAP(Pup_cg[0], Pup_cg[no_cgmms_init_guess]);
+          SWAP(Pdn_cg[0], Pdn_cg[no_cgmms_init_guess]);
```

Comment:
@kostrzewa here I do swaps of the shifts and solution vectors in order to solve a larger one but still including the smaller